src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -57,49 +57,67 @@ public DiagnosticService([ImportMany] IEnumerable<Lazy<IAsynchronousOperationLis
 
         private void RaiseDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs args)
         {
+            Contract.ThrowIfNull(sender);
+            var source = (IDiagnosticUpdateSource)sender;
+
             var ev = _eventMap.GetEventHandlers<EventHandler<DiagnosticsUpdatedArgs>>(DiagnosticsUpdatedEventName);
-            if (ev.HasHandlers)
+            if (!RequireRunningEventTasks(source, ev))
```

Comment:
I will put this in as comment. 

..

basically there are 2 cases when there is no event handler registered. 
first case is when diagnostic source itself provide GetDiagnostics functionality. in that case, DiagnosticService itself doesn't need to track diagnostics reported. in that case, it bail out right away.
second case is when diagnostic source doesn't provide GetDiagnostics functionality. in that case, DiagnosticService itself need to track diagnostics reported. so it need to enqueue background work to process in regardless whether there is event handler registered or not.

this could be separated in 2 tasks, but we already saw case where there are too many tasks enqueued, so I merged it to one.

the reason one might provide GetDiagnostics and one might not is that, if Diagnostic Update source has its own cache since it could have millions of diagnostics (DiagnosticAnalyzer), then it provide GetDiagnostics functionality itself, but if one is like EnC or diagnostics from VS to indicate some issue in VS that expects to have small number of diagnostics, then, it will let DiagnosticService to handle it.